cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.cdap.api.metadata;
+
+import java.util.Map;
+import java.util.Set;
+import javax.ws.rs.NotFoundException;
+
+/**
+ * The context for reading metadata from program
```

Comment:
From which phase of the program this functionality - `reading metadata` would be available? Will this be available in the `initialize` method or in the `prepareLaunch` which happens before provisioning?

Reply:
Didn't we say that we need a way to "announce" in the prepareLaunch() what meta data entities the program will read at runtime? And the runtime will then make all meta data for these entities available?

Reply:
I see. So this interface `MetadataReaderContext` is actually available to the program runtime context such as `initialize` method of the program, right? And only those metadata entities can be used for query which are actually announced in the `prepareLaunch`. Announcing the Metadata entity which will be needed during the runtime will be part of separate interface.

We should probably clearly mention that in the javadoc.